### PR TITLE
Module API - New gates

### PR DESCRIPTION
**Pull request by nmheim**
pyq.modules:
- Bumped torch version to 2.0.0
- Added CZ,CY
- Added MultiParamRotationGate with batched U gate
- Added batched HamEvo, batched HamEvoEig
- Added tests for circuits, gates

### CODE REVIEW

File: pyqtorch/modules/primitive.py [hunk 1]
```diff
@@ -48,14 +76,14 @@ def __init__(self, gate: str, qubits: ArrayLike, n_qubits: int):
...
-        return _apply_gate(state, self.mat, self.qubits, self.n_qubits)
+        return _apply_gate(state, self.matrix, self.qubits, self.n_qubits)
 
     @property
     def device(self) -> torch.device:
-        return self.mat.device
+        return self.matrix.device
 
     def extra_repr(self) -> str:
         return f"'{self.gate}', {self.qubits}, {self.n_qubits}"
```
Comment (madagra) [inline pyqtorch/modules/primitive.py:89]: @nmheim So now it will be extremely easy to add CY and CZ as well here.

**Review by madagra — COMMENTED** [review]
I see a very limited amount of tests for the new operations (subclassing the `torch.nn.Module`). We should likely replace all the previous tests for single gates with the corresponding module-based operation. This can also be done in a separate PR in which we can also switch fully to module-based operations.

File: pyqtorch/modules/parametric.py [hunk 2]
```diff
@@ -20,11 +23,18 @@ def __init__(self, gate: str, qubits: ArrayLike, n_qubits: int, param_name: str)
...
         batch_size = len(theta)
-        mats = rot_matrices(theta, self.paulimat, self.imat, batch_size)
-        return _apply_batch_gate(state, mats, self.qubits, self.n_qubits, batch_size)
+        return rot_matrices(theta, self.paulimat, self.imat, batch_size)
+
+    def apply(self, matrices: torch.Tensor, state: torch.Tensor) -> torch.Tensor:
+        batch_size = matrices.size(-1)
+        return _apply_batch_gate(state, matrices, self.qubits, self.n_qubits, batch_size)
+
+    def forward(self, thetas: dict[str, torch.Tensor], state: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (nmheim) [reply pyqtorch/modules/parametric.py:35]: It removed quite a bit of boilerplate code in other packages and I thought to myself that the module API is so breaking that one more change won't be a problem. its almost like the module API is a completel re-implementation of pyq...

[the remark this answers is not in the document]
Comment (nmheim) [reply pyqtorch/modules/parametric.py:35]: Maybe a nicer way would be to:
- accept only tensors in the gates/operations
- accept only dicts of tensors in the QuantumCircuit

[the remark this answers is not in the document]
Comment (madagra) [reply pyqtorch/modules/parametric.py:35]: Honestly, I would prefer keeping the tensors at least in the gates. Most of the gates accept one or two parameters, so the dictionary is really unnecessary, I think.

[the remark this answers is not in the document]
Comment (nmheim) [reply pyqtorch/modules/parametric.py:35]: From my point of view we need a place where to store which gate gets which parameter. If it’s not happening in the gate itself it should probably happen in the circuit. That would mean that the list of operations would be come a dict of some kind I guess?

[the remark this answers is not in the document]
Comment (madagra) [reply pyqtorch/modules/parametric.py:35]: I think at the circuit level it is acceptable to use the `dict`.

File: pyqtorch/modules/parametric.py [hunk 3]
```diff
@@ -62,3 +97,33 @@ def RY(*args: Any, **kwargs: Any) -> RotationGate:
...
+def CRZ(*args: Any, **kwargs: Any) -> RotationGate:
+    return ControlledRotationGate("Z", *args, **kwargs)
+
+
+class CPHASE(Module):
+    def __init__(self, qubits: ArrayLike, n_qubits: int, param_name: str):
+        super().__init__()
+        self.qubits = qubits
+        self.n_qubits = n_qubits
+        self.param_name = param_name
```
[the remark this answers is not in the document]
Comment (madagra) [reply pyqtorch/modules/parametric.py:119]: OK, this relates to my comment above then.

**Pull request merged**